### PR TITLE
Mysqld.GetSchema: tolerate tables being dropped while inspecting schema

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Identify the two scenarios where a table may be dropped while we're inspecting the schema. Either we get a `(errno 1146)` error code, or the table does not appear in `information_schema` even after it did, before. 

The fix is to silently ignore tables hitting this scenario.

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/12640

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by shlomi-noach** [comment]
hmmmm got this relevant error in `Unit_test`:

```
2023-03-16T13:25:15.6015797Z === RUN   TestDeferSecondaryKeys/3FPK1SK_ShardMerge
2023-03-16T13:25:15.6015939Z === RUN   TestDeferSecondaryKeys/0FPK2tSK
2023-03-16T13:25:15.6016109Z     vreplicator_test.go:443: 
2023-03-16T13:25:15.6016645Z         	Error Trace:	/home/runner/work/vitess/vitess/go/vt/vttablet/tabletmanager/vreplication/vreplicator_test.go:443
2023-03-16T13:25:15.6016885Z         	Error:      	Received unexpected error:
2023-03-16T13:25:15.6017622Z         	            	unexpected number of table definitions returned from GetSchema call for table "t1": 0
2023-03-16T13:25:15.6017881Z         	Test:       	TestDeferSecondaryKeys/0FPK2tSK
2023-03-16T13:25:15.6018020Z === RUN   TestDeferSecondaryKeys/2SKRetryNoErr
2023-03-16T13:25:15.6018181Z === RUN   TestDeferSecondaryKeys/2SKRetryNoErr2
2023-03-16T13:25:15.6018338Z === RUN   TestDeferSecondaryKeys/SKSuperSetNoErr
```

**Comment by shlomi-noach** [comment]
I'm not sure what the situation is in `vreplicator_test.go` calling `stashSecondaryKeys()`, that results with an empty schema. That is, I'm not sure if the new logic is doing the right thing, and this is unexpected to `stashSecondaryKeys()`, or whether the new logic is doing something wrong. I'm not sure if a table is being dropped here.

**Comment by shlomi-noach** [comment]
Ah, I get it. A table without `PRIMARY KEY`. Converting this PR to `draft` while I fix it.

**Comment by shlomi-noach** [comment]
OK, the assumption about the 2nd race condition was incorrect. A table that does not appear in `INFORMATION_SCHEMA.STATISTICS` isn't necessarily dropped. It may also be a table that does not have any keys.

Removed the 2nd race condition handling altogether. It a table does end up being dropped before calculating primary key, it will just have an implicit `nil` primary key.

File: go/vt/mysqlctl/schema.go [hunk 1]
```diff
@@ -109,6 +110,16 @@ func (mysqld *Mysqld) GetSchema(ctx context.Context, dbName string, request *tab
 
 			fields, columns, schema, err := mysqld.collectSchema(ctx, dbName, td.Name, td.Type, request.TableSchemaOnly)
 			if err != nil {
+				// there's a possible race condition: it could happen that a table was dropped in between reading
+				// the list of tables (collectBasicTableData(), earlier) and the point above where we investigate
+				// the table.
+				// This is fine. We identify the situation and continue to remove the table from our records
+				sqlErr, isSQLErr := mysql.NewSQLErrorFromError(err).(*mysql.SQLError)
+				if isSQLErr && sqlErr != nil && sqlErr.Number() == mysql.ERNoSuchTable {
+					td.Fields = nil
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/mysqlctl/schema.go]: OK, I rewrote this in a different way. Problem is that we're iterating `tds` concurrently. Any change to `tds` itself during iteration is less than ideal. So instead I'm building a `validTds` slice which in turn is the final list of table definitions.

File: go/vt/mysqlctl/schema.go [hunk 2]
```diff
@@ -109,6 +110,16 @@ func (mysqld *Mysqld) GetSchema(ctx context.Context, dbName string, request *tab
 
 			fields, columns, schema, err := mysqld.collectSchema(ctx, dbName, td.Name, td.Type, request.TableSchemaOnly)
 			if err != nil {
+				// there's a possible race condition: it could happen that a table was dropped in between reading
+				// the list of tables (collectBasicTableData(), earlier) and the point above where we investigate
+				// the table.
+				// This is fine. We identify the situation and continue to remove the table from our records
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/mysqlctl/schema.go]: fixed.

File: go/vt/mysqlctl/schema.go [hunk 3]
```diff
@@ -146,7 +159,14 @@ func (mysqld *Mysqld) GetSchema(ctx context.Context, dbName string, request *tab
 		td.PrimaryKeyColumns = colMap[td.Name]
 	}
 
-	sd.TableDefinitions = tds
+	sd.TableDefinitions = []*tabletmanagerdatapb.TableDefinition{}
+	for _, td := range tds {
+		if td.Fields == nil {
+			// means table was dropped before we called GetColumns
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/mysqlctl/schema.go]: fixed.

File: go/vt/mysqlctl/schema.go [hunk 4]
```diff
@@ -109,6 +112,15 @@ func (mysqld *Mysqld) GetSchema(ctx context.Context, dbName string, request *tab
 
 			fields, columns, schema, err := mysqld.collectSchema(ctx, dbName, td.Name, td.Type, request.TableSchemaOnly)
 			if err != nil {
+				// there's a possible race condition: it could happen that a table was dropped in between reading
```
Comment (mattlord) [inline go/vt/mysqlctl/schema.go:115]: IF you have to do another push for some reason, we can capitalize `There's` here. It's annoyingly nitty, I know, so no need unless we have to do another commit/push for another reason.

**Comment by shlomi-noach** [comment]
> IMO, we should re-run the affected test(s) 5-10 times before merging to be sure 

I've ran this on local env some 20 times and it passed all runs. I'm signing out for the week and won't be watching CI to kick restarts. If this can wait until Sunday, I can verify and merge then. If not, anyone should feel free to test/retry and merge.

**Comment by shlomi-noach** [comment]
Multiple tests are consistently failing on this PR. Example: `vreplication_basic`: https://github.com/vitessio/vitess/actions/runs/4438816245/jobs/7831448081?pr=12641 with error:

```
2023-03-19T05:54:35.2013877Z Reshard Error: rpc error: code = Unknown desc = copySchema: CopySchemaShard was not successful because the schemas between the two tablets cell:"zone1" uid:200 and cell:"zone1" uid:600 differ: [dest has an extra table named cproduct dest has an extra table named customer dest has an extra table named db_order_test source has an extra table named db_order_test source has an extra table named cproduct source has an extra table named customer]
```

`vtgate_schema` has a similar error:

```
Mar 19 05:54:00         	Error Trace:	go/test/endtoend/vtgate/schema/schema_test.go:249
Mar 19 05:54:00         	            				go/test/endtoend/vtgate/schema/schema_test.go:110
Mar 19 05:54:00         	Error:      	Expected nil, but got: &exec.ExitError{ProcessState:(*os.ProcessState)(0xc0006de408), Stderr:[]uint8(nil)}
Mar 19 05:54:00         	Test:       	TestSchemaChange
```

I'm going to try a different angle to this PR. Instead of removing the dropped table from the result set, I'm going to keep it there, with empty table definition, ie empty keys, empty columns etc.

**Comment by shlomi-noach** [comment]
@mattlord @rohit-nayak-ps could you give this another look? I've simplified the change to just returning an incomplete table description. Previously, when I excluded the table description from the result table list, many tests failed, consistently, see above. Now, everything seems to pass.

**Comment by shlomi-noach** [comment]
PS I'm re-running `onlineddl_vrepl` in CI just to ensure it passes consistently.

**Comment by shlomi-noach** [comment]
So with the current implementation `Columns` and `Fields` remain `nil`. I reviewed all usage and nothing is susceptible to nil pointer access; everything just properly `range`s over the slices.

Tests are super happy. I think this should be good.

**Comment by rohit-nayak-ps** [comment]
> @mattlord @rohit-nayak-ps could you give this another look? I've simplified the change to just returning an incomplete table description. Previously, when I excluded the table description from the result table list, many tests failed, consistently, see above. Now, everything seems to pass.

Looks good @shlomi-noach, thanks!

**Comment by shlomi-noach** [comment]
waiting to see if the bot is able to cherry pick this
